In [11]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn import linear_model

In [12]:
data = pd.read_csv('Weather.csv')

In [13]:
data['DidItRain'] = data['RainTotal'].apply(lambda x: 1 if x > 0 else 0)

In [14]:
data

Date  TempMax  TempAvg  TempMin  DewMax  DewAvg  DewMin  HumidityMax  \
0       1117       19     14.6        2      14    10.5      -1           92   
1       1217       23     16.2       14      17    12.8      11           92   
2       1317       24     21.0       17      21    18.0      14           96   
3       1417       26     23.6       20      23    21.3      17           96   
4       1517       21     11.7       -4      18     6.5     -10           95   
...      ...      ...      ...      ...     ...     ...     ...          ...   
1090  122719       35     31.0       27      25    23.3      21           85   
1091  122819       38     30.5       25      26    24.0      22           92   
1092  122919       32     30.6       28      30    26.4      22           92   
1093  123019       34     31.5       29      26    23.8      22           82   
1094  123119       40     36.4       32      33    27.7      23           79   

      HumidityAvg  HumidityMin  WindMax  WindAvg  WindMin  PressureMax  \
0            84.1           71        9      3.0        0         26.9   
1            86.7           78        8      2.7        0         27.1   
2            88.5           71        6      1.4        0         27.1   
3            90.8           85       10      2.6        0         27.1   
4            80.0           64       12      5.3        0         27.3   
...           ...          ...      ...      ...      ...          ...   
1090         73.5           59        9      5.7        0         27.2   
1091         77.9           57       12      2.1        0         27.2   
1092         84.5           75       10      5.6        0         27.2   
1093         73.5           61        7      2.2        0         27.4   
1094         71.1           63       18     10.2        0         27.4   

      PressureAvg  PressureMin  RainTotal  DidItRain  
0            26.8         26.7       0.00          0  
1            26.9         26.8       0.03          1  
2            27.1         27.0       0.02          1  
3            27.0         26.9       0.25          1  
4            27.2         27.1       0.34          1  
...           ...          ...        ...        ...  
1090         27.1         27.1       0.00          0  
1091         27.2         27.2       0.00          0  
1092         27.1         27.1       0.00          0  
1093         27.3         27.2       0.05          1  
1094         27.2         27.0       0.00          0  

[1095 rows x 18 columns]

In [35]:
feature_cols = ['TempAvg','DewAvg','HumidityAvg','WindAvg','PressureAvg'] 
#decided to do averages since someone would probably know those features better than the max/min for the day
X = data[feature_cols] # Features
y = data.DidItRain # Target variable

In [36]:
X_train,X_test,y_train,y_test=train_test_split(X, y, test_size=0.25, random_state=25)

In [37]:
# instantiate the model (using the default parameters)
logreg = LogisticRegression(max_iter=5000)

# fit the model with data
logreg.fit(X_train,y_train)


y_pred = logreg.predict(X_test)

/Applications/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [38]:
X_test

TempAvg  DewAvg  HumidityAvg  WindAvg  PressureAvg
330     44.2    31.3         61.5     11.7         27.0
373     42.5    39.9         90.6      7.6         26.7
799     37.5    21.5         56.3      3.7         27.2
279     56.3    32.8         42.7     10.6         26.9
174     72.5    37.2         31.3      5.6         27.1
..       ...     ...          ...      ...          ...
830     45.3    36.0         70.8      8.6         27.0
748     39.6    18.2         43.8     12.4         27.2
572     83.7    38.0         22.0      6.8         27.0
994     63.8    40.3         45.9      8.5         27.0
717     41.4    31.0         67.6      4.3         27.3

[274 rows x 5 columns]

In [39]:
print("Accuracy:",accuracy_score(y_test, y_pred))

Accuracy: 0.8175182481751825


In [41]:
#Prediction
#['Date','TempAvg','DewAvg','HumidityAvg','WindAvg','PressureAvg']
new_date = [[51.9, 40, 75.2, 2.5, 26.9]]
#Checked an actual date that was in the test set that was true for rain but set it to 2020 and changed a few other things
prediction = logreg.predict(new_date)
prediction[0]

1

## Findings
### Is it accurate?
I think this model is pretty accurate. It is not the best but 81% accuracy would still be better than taking a blind guess. Out of all my data it rained 270 days in 3 years which is 24% of all days. This logistic model predicts the accuracy of it raining much better. If we were to not blind guess we would probably have a better chance since we can choose months where rain is more likley like March or April but since it is a blind guess we can assume that our model is better at predicting.